apps/os/backend/durable-objects/estate-builds.ts

```diff
@@ -0,0 +1,445 @@
+      this.watchersByBuildId.set(buildId, watchers);
+    });
+  }
+
+  async fetch(request: Request): Promise<Response> {
+    const url = new URL(request.url);
+
+    // Lazy initialize storage (tables etc.)
+    await this.ensureInitialized();
+
+    if (request.headers.get("Upgrade") === "websocket") {
+      return this.handleWebSocketUpgrade(request);
+    }
+
+    if (url.pathname === "/ingest" && request.method === "POST") {
+      return this.handleIngest(request);
+    }
+
+    if (url.pathname === "/logs" && request.method === "GET") {
+      return this.handleGetLogs(request);
+    }
+
+    return new Response("Not found", { status: 404 });
+  }
+
+  private async handleIngest(request: Request): Promise<Response> {
+    try {
+      const url = new URL(request.url);
+      const buildId = url.searchParams.get("buildId");
+      const estateId = url.searchParams.get("estateId");
+      if (!buildId) return new Response("Missing buildId", { status: 400 });
+      if (!estateId) return new Response("Missing estateId", { status: 400 });
+
+      const body = (await request.json()) as unknown;
+      const rawLogs = (body && typeof body === "object" && (body as any).logs) || [];
+      const logs: Array<{
+        seq: number;
+        ts: number;
+        stream: "stdout" | "stderr";
+        message: string;
+        event?: "BUILD_STARTED" | "BUILD_SUCCEEDED" | "BUILD_FAILED" | "CONFIG_OUTPUT";
+      }> = Array.isArray(rawLogs) ? rawLogs : [];
+
+      // Determine last processed sequence from SQL (no KV)
+      let lastSeq = 0;
+      try {
+        const rows = await this.ctx.storage.sql.exec(
+          "SELECT MAX(seq) AS max_seq FROM logs WHERE build_id = ?",
+          buildId,
+        );
+        if (Array.isArray(rows) && rows.length > 0) {
+          const maxSeqVal = (rows[0] as any)?.max_seq;
+          if (maxSeqVal !== null && maxSeqVal !== undefined) lastSeq = Number(maxSeqVal) || 0;
+        }
+      } catch (_err) {
+        // default to 0
+      }
+
+      // Process in ascending seq order and insert only newer entries
+      logs.sort((a, b) => a.seq - b.seq);
+      let newLast = lastSeq;
+      let processedAny = false;
+      let status: "none" | "in_progress" | "complete" | "failed" = "none";
+      for (const entry of logs
+        .filter((e) => typeof e.seq === "number" && e.seq > lastSeq)
+        .sort((a, b) => a.seq - b.seq)) {
+        const ts = Number(entry.ts) || Date.now();
+        const stream = entry.stream === "stderr" ? "stderr" : "stdout";
+        const message = String(entry.message ?? "");
+        try {
+          await this.ctx.storage.sql.exec(
+            "INSERT INTO logs (build_id, seq, ts, stream, message, event) VALUES (?, ?, ?, ?, ?, ?)",
+            buildId,
+            entry.seq,
+            ts,
+            stream,
+            message,
+            entry.event ?? null,
+          );
+        } catch (_err) {
+          // Fallback for older schema without seq/event columns
```

Comment:
do we have to? can't we just have simpler code and a good experience for _future_ builds?